File: DFS_BFS/[영선] DFS_BFS/네트워크.swift
//
//  네트워크.swift
//  Algorithm
//
//  Created by 임영선 on 2023/10/13.
//

import Foundation

/*
 이어져있는 덩어리 개수를 구하는 문제다.
 
 1. DFS로 쭉쭉 갈 수 있는 곳까지 돌린다.
 2. 포문으로 버텍스 첨부터 끝까지 돌리고 방문안한 곳 또 DFS 돌린다.
 3. DFS가 모두 빠져나온 순간 덩어리 하나로 취급해서 +1한다.
 
 */

func solution(_ n:Int, _ computers:[[Int]]) -> Int {
    var isVisited = Array(repeating: false, count: n)
    var answer = 0
    
    func DFS(_ v: Int) {
        for i in 0..<n {
            // 방문하지 않았고 이어져있으면(1이면)
            if !isVisited[i] && computers[v][i] == 1 {
                isVisited[i] = true
                DFS(i)
            }
        }
    }
    
    for i in 0..<n {
        if !isVisited[i] {
            DFS(i)
            answer += 1
        }
    }
    
    return answer
    
}

File: DFS_BFS/[희진]DFS_BFS/네트워크.js
function solution(n, computers) {
  // 방문 여부를 저장하는 배열
  const visited = [];
  // 네트워크 개수를 저장하는 변수
  let answer = 0;

  // 깊이 우선 탐색 함수
  const dfs = (node) => {
    // 현재 노드를 방문 표시
    visited[node] = true;

    // 현재 노드와 연결된 다른 노드들에 대해 재귀적으로 방문
    for (let i = 0; i < n; i++) {
      if (computers[node][i] === 1 && !visited[i]) {
        dfs(i);
      }
    }
  };

  // 모든 노드에 대해 반복
  for (let i = 0; i < n; i++) {
    // 방문하지 않은 노드일 경우에만 DFS 수행하고 네트워크 개수 증가
    if (!visited[i]) {
      dfs(i);
      answer++;
    }
  }

  return answer;
}
